src/util.cc

```diff
@@ -192,22 +194,77 @@ void CanonicalizePath(char* path, size_t* len, uint64_t* slash_bits) {
       continue;
     }
 
-    if (component_count == kMaxPathComponents)
-      Fatal("path has too many components : %s", path);
     components[component_count] = dst;
-    ++component_count;
+    if (++component_count > kMaxPathComponents)
+      Fatal("path has too many components : %s", path);
+
+      // Do no compile regular loop code with GCC (see note below).
+#if !(defined(__GNUC__) && !defined(__clang__))
+    for (;;) {
+      *dst++ = *src++;
+      if (src == end)
+        break;
+      if (IsPathSeparator(*src)) {
+        *dst++ = *src++;  // Copy final path separator if any.
+        break;
+      }
+    }
+#else  // GCC (but not Clang)
+    // The following code path is used with GCC because the above loop
+    // works well with `gcc-11 -O2`, but results in disastrous code with
+    // `gcc-11 -O3`, for some unknown reason. Comparisons of canon_perftest
+    // results (minimal times of 5 runs each):
+    //
+    //   compiler              regular      unrolled
+    //
+    //   clang-13 -O2            114ms        114ms
+    //   clang-13 -O3            110ms        110ms
+    //   gcc-11 -O2              126ms        110ms
+    //   gcc-11 -O3              275ms !!     110ms
```

Comment:
IMHO, it would be better to report this as a bug to GCC and keep the less complex code.